presto-main/src/main/java/com/facebook/presto/util/PeriodicTaskExecutor.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.util;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.LongUnaryOperator;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Math.round;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class PeriodicTaskExecutor
+        implements AutoCloseable
+{
+    private final long delayTargetMillis;
+    private final ScheduledExecutorService executor;
+    private final Runnable runnable;
+    private final LongUnaryOperator nextDelayFunction;
+
+    @GuardedBy("this")
+    private long delayMillis;
+    @GuardedBy("this")
+    private ScheduledFuture<?> scheduledFuture;
+    @GuardedBy("this")
+    private boolean stopped;
+
+    public PeriodicTaskExecutor(long delayTargetMillis, ScheduledExecutorService executor, Runnable runnable)
+    {
+        this(delayTargetMillis, executor, runnable, PeriodicTaskExecutor::nextDelayWithJitterMillis);
+    }
+
+    public PeriodicTaskExecutor(long delayTargetMillis, ScheduledExecutorService executor, Runnable runnable, LongUnaryOperator nextDelayFunction)
+    {
+        checkArgument(delayTargetMillis > 0, "delayTargetMillis must be > 0");
+        this.delayTargetMillis = delayTargetMillis;
+        this.executor = requireNonNull(executor, "executor is null");
+        this.runnable = requireNonNull(runnable, "runnable is null");
+        this.nextDelayFunction = requireNonNull(nextDelayFunction, "nextDelayFunction is null");
+        tick();
+    }
+
+    private synchronized void tick()
+    {
+        scheduledFuture = executor.schedule(this::run, delayMillis, MILLISECONDS);
+    }
+
+    private synchronized void run()
+    {
+        forceRun();
+        delayMillis = nextDelayFunction.applyAsLong(delayTargetMillis);
+        if (!stopped) {
```

Comment:
do we want to also check if the previous request is finished or not before sending a new one?

Reply:
It will only hit this code unless the previous request has finished.